src/Avalonia.Base/PropertyStore/EffectiveValue`1.cs

```diff
@@ -0,0 +1,260 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using Avalonia.Data;
+
+namespace Avalonia.PropertyStore
+{
+    /// <summary>
+    /// Represents the active value for a property in a <see cref="ValueStore"/>.
+    /// </summary>
+    /// <remarks>
+    /// Stores the active value in an <see cref="AvaloniaObject"/>'s <see cref="ValueStore"/>
+    /// for a single property, when the value is not inherited or unset/default.
+    /// </remarks>
+    internal sealed class EffectiveValue<T> : EffectiveValue
+    {
+        private T? _baseValue;
+        private UncommonFields? _uncommon;
+
+        public EffectiveValue(
+            AvaloniaObject owner,
+            StyledPropertyBase<T> property,
+            T value, 
+            BindingPriority priority)
+        {
+            Value = value;
+            Priority = priority;
+            BasePriority = BindingPriority.Unset;
+
+            if (property.HasCoercion &&
```

Comment:
Ok, implemented this. I also use the cached metadata when inherited values change now, which may have different behavior in the case of inherited values which have overridden default values. I think we need to disallow overriding default values for inherited properties because the semantics aren't clear anyway.